Expose zkey to martian_atmosphere_api, which defines the altitude of …

gsc10 opened the pull request:
…the profile

The current default is zkey=2, which implies the altitude is the altitude above the Martian zero datum (Mars geoid or “areoid”/MOLA_0) (km). However, for some applications the altitude above local surface (zkey=3) is preferable. I have exposed them all here, in case we ever need them, but 1, 4 and 5 are not appropriate for the current fcm.

gsc10 committed:
Update crater_tools.py

gsc10 committed:
Expose zkey to martian_atmosphere_api, which defines the altitude of the profile

The current default is zkey=2, which implies the altitude is the altitude above the Martian zero datum (Mars geoid or “areoid”/MOLA_0) (km). However, for some applications the altitude above local surface (zkey=3) is preferable. I have exposed them all here, in case we ever need them, but 1, 4 and 5 are not appropriate for the current fcm.

gsc10 committed:
Add dry soil ground type

gsc10 committed:
Fix URL for MCD query to be consistent with manual query

gsc10 committed:
Update _mars_atm_api.py

acse-ds2419 commented on fcm/atmosphere/_mars_atm_api.py:
```diff
@@ -32,41 +32,59 @@ def martian_atmosphere_api(latitude, longitude, timestamp):
         timestamp for which to request the data
-    
+
+    zkey : int
```
Missing type hint here that it's an optional parameter:
```python
zkey : int, optional
```

gsc10 replied:
done

acse-ds2419 commented on fcm/atmosphere/_mars_atm_api.py:
```diff
@@ -32,41 +32,59 @@ def martian_atmosphere_api(latitude, longitude, timestamp):
+    # MCD provides results in m (or Pa); convert to km (kPa) and
+    # set appropriate index name
+    dataframe.index *= 1e-3 
+    dataframe.index.name = altitude_type[zkey]
```
You should include a check at the beginning of the function that `zkey` is actually an integer between 1 and 5. Otherwise this will only fail here with a key error, after the http request has already been made. Or fail with a cryptic http error because the API doesn't understand whatever the user put as `zkey`. You can use my helper function:
```python3
zkey = _check_number(zkey, "zkey", False, 1, True, 5, True)
```

gsc10 replied:
done

acse-ds2419 commented on fcm/atmosphere/_mars_atm_api.py:
```diff
@@ -32,41 +32,59 @@ def martian_atmosphere_api(latitude, longitude, timestamp):
         jdate += timestamp.hour / 24 + timestamp.minute / (24*60) + timestamp.second / (24*3600)
     
-    url = BASE_URL + "cgi-bin/mcdcgi.py?"
+    url = BASE_URL + "cgi-bin/mcdcgi.py?&datekeyhtml=0&localtime=0"
```
What are these two new parameters doing?

gsc10 replied:
I think these are necessary so that `julian` is correctly interpreted. Without these parameters, the date was being incorrectly interpreted (but I am not sure how). See http://www-mars.lmd.jussieu.fr/mars/info_web/user_manual_5.3.pdf

acse-ds2419 commented on fcm/atmosphere/_mars_atm_api.py:
```diff
@@ -32,41 +32,59 @@ def martian_atmosphere_api(latitude, longitude, timestamp):
     url += "&julian={:.5f}&latitude={:.9f}&longitude={:.9f}".format(jdate, latitude, longitude)
-    url += "&altitude=all&zkey=2&var1=rho&colorm=jet"
-    
+    url += "&altitude=all&zonmean=off&hrkey=1&zkey="+str(zkey)+"&var1=rho&colorm=jet"
```
use integer formatting here:
```python3
url += "&altitude=all&zonmean=off&hrkey=1&zkey={:.d}&var1=rho&colorm=jet".format(zkey)
```

Also same question about the two other additional parameters `znonmean` and `hrkey`

gsc10 replied:
On the web-based interface, these are the "Use high-resolution topography" and "Zonal averaging" switches.

acse-ds2419 commented on fcm/atmosphere/_mars_atm_api.py:
```diff
@@ -102,7 +120,7 @@ def _load_and_parse_txt_file(url, timeout=2):
         "expected only two columns, got {:d}".format(atmosphere.shape[1] + 1)
     atmosphere.columns = ["Density (kg/m3)"]
-    atmosphere.index.name = "altitude above MOLA_0 (m)"
+    atmosphere.index.name = "altitude"
```
Looks to me like this line was redundant before, and still is, as it's overwritten on line 70 anyway. => should be removed

gsc10 replied:
Yup, done.

acse-ds2419 reviewed (changes requested):
Looks good, some very small tweaks are necessary (most importantly check that the new parameter `zkey` is indeed an int in {1, ... , 5})

gsc10 committed:
Add check for valid zkey input and tidy up

Thanks Dominic!

acse-ds2419 committed:
added venv/ folder for the project's virtual env to .gitignore

acse-ds2419 committed:
fixed broken unit test

acse-ds2419 committed:
- zonmean=off breaks things,
  since then the web service
  returns two 2D tables
- using the `params` arg for requests.get
  instead of composing URL directly

acse-ds2419 committed:
Merge branch 'master' into gsc-MCD-api-improvement

acse-ds2419 commented on fcm/atmosphere/_mars_atm_api.py:
```diff
@@ -32,51 +32,87 @@ def martian_atmosphere_api(latitude, longitude, timestamp):
+              'latitude': f"{latitude:.9f}",
+              'longitude': f"{longitude:.9f}",
+              'altitude': "all",
+              'averaging': "off",  # turns off zonal averaging
```
I tested your additional parameters in the URL, and setting `zonmean=off` has the effect that the web service returns two 2D tables instead of one 1D table like we want. After poking around a bit I found the parameter `averaging`, which at least doesn't seem to break things. @gsc10 what do you think?

acse-ds2419 commented:
Switched to using the `params` parameter of `requests.get()`, which makes things quite a bit cleaner

gsc10 committed:
Update crater_tools.py

gsc10 committed:
Merge branch 'gsc-SFM-clusters' into gsc-MCD-api-improvement
